classes/core/PKPPageRouter.php

```diff
@@ -522,14 +521,33 @@ private function _setLocale(PKPRequest $request, ?string $setLocale): void
             $request->setCookieVar('currentLocale', $newLocale);
         }
 
-        // Do not permit basic auth strings (user:password@url) in source parameter for redirects
-        if (preg_match('#^/\w#', $source = str_replace('@', '', $request->getUserVar('source') ?? ''))) {
-            $request->redirectUrl($source);
+        $indexUrl = $this->getIndexUrl($request);
+        if (isset($setLocale)) {
+            $targetUrl = '';
+            // The source parameter is coming either from the languageToggle block plugin
+            // and contains $smarty.server.SERVER_NAME|cat:$smarty.server.REQUEST_URI
+            // i.e. the whole URL except the protocol,
+            // or from TopNavActions and contain the whole URL with protocol
+            $source = $request->getUserVar('source');
+            if (isset($source)) {
+                $targetUrl = str_replace('@', '', $source);
+                $protocol = $request->getProtocol();
+                if (!str_starts_with($source, $protocol)) {
+                    $targetUrl = $request->getProtocol() . '://' . $targetUrl;
+                }
+            } elseif (isset($_SERVER['HTTP_REFERER'])) {
+                $targetUrl = $_SERVER['HTTP_REFERER'];
```

Comment:
I think the `HTTP_REFERER` can be whatever the requester specifies, so this could potentially be a dangerous URL (e.g. something that goes to another domain). I think the only way we can prevent this from being done is to match the `allowed_hosts`.